pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,487 @@
+		}
+	}
+}
+
+func (c *Cluster) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Cluster) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect strings (ExternalIPPool name) to come off the work queue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the work queue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncConsistentHash(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the work queue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing consistentHash by ExternalIPPool failed, requeue", "ExternalIPPool", key)
+	}
+	return true
+}
+
+func (c *Cluster) syncConsistentHash(eipName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing consistentHash", "ExternalIPPool", eipName, "durationTime", time.Since(startTime))
+	}()
+
+	eip, err := c.externalIPPoolLister.Get(eipName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			c.consistentHashRWMutex.Lock()
+			defer c.consistentHashRWMutex.Unlock()
+			delete(c.consistentHashMap, eipName)
+			return nil
+		}
+		return err
+	}
+
+	// updateConsistentHash refresh the consistentHashMap.
+	updateConsistentHash := func(eip *v1alpha2.ExternalIPPool) error {
+		nodeSel, err := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+		if err != nil {
+			return err
+		}
+		nodes, err := c.nodeLister.List(nodeSel)
+		if err != nil {
+			return err
+		}
+		aliveNodes := c.aliveNodes()
+		// Node alive and Node labels matches with ExternalIPPool nodeSelector.
+		var aliveAndMatchedNodes []string
+		for _, node := range nodes {
+			nodeName := node.Name
+			if aliveNodes.Has(nodeName) {
+				aliveAndMatchedNodes = append(aliveAndMatchedNodes, nodeName)
+			}
+		}
+		consistentHashMap := newNodeConsistentHashMap()
+		consistentHashMap.Add(aliveAndMatchedNodes...)
+		c.consistentHashRWMutex.Lock()
+		defer c.consistentHashRWMutex.Unlock()
+		c.consistentHashMap[eip.Name] = consistentHashMap
+		c.notify(eip.Name)
+		return nil
+	}
+
+	if err := updateConsistentHash(eip); err != nil {
+		return err
+	}
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+func (c *Cluster) handleClusterNodeEvents(nodeEvent *memberlist.NodeEvent) {
+	node, event := nodeEvent.Node, nodeEvent.Event
+	switch event {
+	case memberlist.NodeJoin, memberlist.NodeLeave:
+		// When a Node joins cluster, all matched ExternalIPPools consistentHash should be updated.
```

Comment:
If you do not like to capitalize "when" in the next sentence, probably change "." to ";"

Reply:
ditto